File: src/main/java/com/flexpoker/table/query/handlers/PlayerRaisedEventHandler.java
package com.flexpoker.table.query.handlers;

import java.util.stream.Collectors;

import javax.inject.Inject;

import org.springframework.stereotype.Component;

import com.flexpoker.chat.service.ChatService;
import com.flexpoker.framework.event.EventHandler;
import com.flexpoker.framework.pushnotifier.PushNotificationPublisher;
import com.flexpoker.login.repository.LoginRepository;
import com.flexpoker.pushnotifications.TableUpdatedPushNotification;
import com.flexpoker.table.command.events.PlayerRaisedEvent;
import com.flexpoker.table.query.repository.TableRepository;
import com.flexpoker.web.dto.outgoing.SeatDTO;
import com.flexpoker.web.dto.outgoing.TableDTO;

@Component
public class PlayerRaisedEventHandler implements EventHandler<PlayerRaisedEvent> {

    private final LoginRepository loginRepository;

    private final TableRepository tableRepository;

    private final PushNotificationPublisher pushNotificationPublisher;

    private final ChatService chatService;

    @Inject
    public PlayerRaisedEventHandler(
            LoginRepository loginRepository,
            TableRepository tableRepository,
            PushNotificationPublisher pushNotificationPublisher,
            ChatService chatService) {
        this.loginRepository = loginRepository;
        this.tableRepository = tableRepository;
        this.pushNotificationPublisher = pushNotificationPublisher;
        this.chatService = chatService;
    }

    @Override
    public void handle(PlayerRaisedEvent event) {
        handleUpdatingTable(event);
        handlePushNotifications(event);
        handleChat(event);
    }

    private void handleUpdatingTable(PlayerRaisedEvent event) {
        var currentTable = tableRepository.fetchById(event.getAggregateId());
        var username = loginRepository.fetchUsernameByAggregateId(event.getPlayerId());

        var raiseToAmount = event.getRaiseToAmount();

        var updatedSeats = currentTable.getSeats().stream()
                .map(seatDTO -> {
                    var amountOverChipsInFront = raiseToAmount - seatDTO.getChipsInFront();

                    if (seatDTO.getName().equals(username)) {
                        var updatedChipsInBack = seatDTO.getChipsInBack() - amountOverChipsInFront;
                        return new SeatDTO(seatDTO.getPosition(),
                                seatDTO.getName(), updatedChipsInBack, raiseToAmount,
                                seatDTO.isStillInHand(), 0, 0, seatDTO.isButton(),
                                seatDTO.isSmallBlind(), seatDTO.isBigBlind(), false);
                    }
                    return new SeatDTO(seatDTO.getPosition(), seatDTO.getName(), seatDTO.getChipsInBack(),
                            seatDTO.getChipsInFront(), seatDTO.isStillInHand(),
                            Math.min(raiseToAmount * 2, seatDTO.getChipsInFront() + seatDTO.getChipsInBack()),
                            amountOverChipsInFront, seatDTO.isButton(), seatDTO.isSmallBlind(), seatDTO.isBigBlind(),
                            seatDTO.isActionOn());
                }).collect(Collectors.toList());

        var previousChipsInFront = currentTable.getSeats().stream()
                .filter(x -> x.getName().equals(username))
                .findAny().get()
                .getChipsInFront();
        var totalPotIncrease = raiseToAmount - previousChipsInFront;

        var updatedTable = new TableDTO(currentTable.getId(),
                event.getVersion(), updatedSeats, currentTable.getTotalPot() + totalPotIncrease,
                currentTable.getPots(), currentTable.getVisibleCommonCards(),
                currentTable.getCurrentHandMinRaiseToAmount(), currentTable.getCurrentHandId());
        tableRepository.save(updatedTable);
    }

    private void handlePushNotifications(PlayerRaisedEvent event) {
        var pushNotification = new TableUpdatedPushNotification(event.getGameId(), event.getAggregateId());
        pushNotificationPublisher.publish(pushNotification);
    }

    private void handleChat(PlayerRaisedEvent event) {
        var username = loginRepository.fetchUsernameByAggregateId(event.getPlayerId());
        var message = username + " raises to " + event.getRaiseToAmount();
        chatService.saveAndPushSystemTableChatMessage(event.getGameId(), event.getAggregateId(), message);
    }

}

File: src/main/kotlin/com/flexpoker/table/query/handlers/PlayerRaisedEventHandler.kt
package com.flexpoker.table.query.handlers

import com.flexpoker.chat.service.ChatService
import com.flexpoker.framework.event.EventHandler
import com.flexpoker.framework.pushnotifier.PushNotificationPublisher
import com.flexpoker.login.repository.LoginRepository
import com.flexpoker.pushnotifications.TableUpdatedPushNotification
import com.flexpoker.table.command.events.PlayerRaisedEvent
import com.flexpoker.table.query.repository.TableRepository
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class PlayerRaisedEventHandler @Inject constructor(
    private val loginRepository: LoginRepository,
    private val tableRepository: TableRepository,
    private val pushNotificationPublisher: PushNotificationPublisher,
    private val chatService: ChatService
) : EventHandler<PlayerRaisedEvent> {

    @Async
    override fun handle(event: PlayerRaisedEvent) {
        handleUpdatingTable(event)
        handlePushNotifications(event)
        handleChat(event)
    }

    private fun handleUpdatingTable(event: PlayerRaisedEvent) {
        val tableDTO = tableRepository.fetchById(event.aggregateId)
        val username = loginRepository.fetchUsernameByAggregateId(event.playerId)
        val raiseToAmount = event.raiseToAmount
        val updatedSeats = tableDTO.seats!!
            .map {
                val amountOverChipsInFront = raiseToAmount - it.chipsInFront
                if (it.name == username) {
                    val updatedChipsInBack = it.chipsInBack - amountOverChipsInFront
                    it.copy(chipsInBack = updatedChipsInBack, chipsInFront = raiseToAmount,
                        raiseTo = 0, callAmount = 0, isActionOn = false)
                } else {
                    it.copy(raiseTo = minOf(raiseToAmount * 2, it.chipsInFront + it.chipsInBack),
                        callAmount = amountOverChipsInFront)
                }
            }
        val previousChipsInFront = tableDTO.seats.first { it.name == username }.chipsInFront
        val totalPotIncrease = raiseToAmount - previousChipsInFront
        val updatedTable = tableDTO.copy(version = event.version, seats = updatedSeats,
            totalPot = tableDTO.totalPot + totalPotIncrease)
        tableRepository.save(updatedTable)
    }

    private fun handlePushNotifications(event: PlayerRaisedEvent) {
        val pushNotification = TableUpdatedPushNotification(event.gameId, event.aggregateId)
        pushNotificationPublisher.publish(pushNotification)
    }

    private fun handleChat(event: PlayerRaisedEvent) {
        val username = loginRepository.fetchUsernameByAggregateId(event.playerId)
        val message = username + " raises to " + event.raiseToAmount
        chatService.saveAndPushSystemTableChatMessage(event.gameId, event.aggregateId, message)
    }

}